File: frontend/lib/schemas/patientExcel.ts
import { z } from 'zod';

export const patientExcelChangeSchema = z.object({
  field: z.string(),
  old_value: z.unknown().nullable(),
  new_value: z.unknown().nullable(),
});

export const patientExcelImportRowSchema = z.object({
  row_number: z.number().int(),
  patient_id: z.string().uuid().nullable(),
  patient_code: z.string().nullable(),
  operation: z.enum(['new', 'update', 'delete', 'error', 'noop']),
  changes: z.array(patientExcelChangeSchema).default([]),
  error_message: z.string().nullable(),
});

export const pfvExcelImportRowSchema = z.object({
  row_number: z.number().int(),
  patient_id: z.string().uuid().nullable(),
  patient_code: z.string().nullable(),
  weekday: z.number().int().nullable(),
  slot_index: z.number().int().nullable(),
  operation: z.enum(['new', 'update', 'delete', 'error', 'noop']),
  changes: z.array(patientExcelChangeSchema).default([]),
  error_message: z.string().nullable(),
});

export const patientExcelImportSummarySchema = z.object({
  patients_new: z.number().int().default(0),
  patients_update: z.number().int().default(0),
  patients_delete: z.number().int().default(0),
  patients_error: z.number().int().default(0),
  patients_noop: z.number().int().default(0),
  pfv_new: z.number().int().default(0),
  pfv_update: z.number().int().default(0),
  pfv_delete: z.number().int().default(0),
  pfv_error: z.number().int().default(0),
  pfv_noop: z.number().int().default(0),
});

export const patientExcelImportResponseSchema = z.object({
  summary: patientExcelImportSummarySchema,
  patient_rows: z.array(patientExcelImportRowSchema),
  pfv_rows: z.array(pfvExcelImportRowSchema),
  transaction_applied: z.boolean(),
});

export type PatientExcelImportResponse = z.infer<typeof patientExcelImportResponseSchema>;
export type PatientExcelImportRow = z.infer<typeof patientExcelImportRowSchema>;
export type PfvExcelImportRow = z.infer<typeof pfvExcelImportRowSchema>;

// ---------------------------------------------------------------------------
// 完全置換 (バックアップ復元) 用 schema
// ---------------------------------------------------------------------------

export const patientExcelReplaceAllSummarySchema = z.object({
  patients_to_create: z.number().int().default(0),
  patients_to_update: z.number().int().default(0),
  patients_to_soft_delete: z.number().int().default(0),
  patients_error: z.number().int().default(0),
  pfv_to_replace: z.number().int().default(0),
  pfv_to_create: z.number().int().default(0),
  pfv_error: z.number().int().default(0),
});

export const patientSoftDeletePreviewSchema = z.object({
  patient_id: z.string(),
  patient_code: z.string().nullable(),
  name: z.string(),
});

export const patientExcelReplaceAllResponseSchema = z.object({
  summary: patientExcelReplaceAllSummarySchema,
  patient_rows: z.array(patientExcelImportRowSchema),
  pfv_rows: z.array(pfvExcelImportRowSchema),
  transaction_applied: z.boolean(),
  patients_to_soft_delete_preview: z.array(patientSoftDeletePreviewSchema).default([]),
});

export type PatientExcelReplaceAllResponse = z.infer<typeof patientExcelReplaceAllResponseSchema>;
export type PatientSoftDeletePreview = z.infer<typeof patientSoftDeletePreviewSchema>;

File: backend/app/schemas/v2/patient_excel.py
"""Pydantic schemas for /api/v1/patients/import-export.

Frontend と契約する Response 型. dry_run / apply で同じ shape を返す.
"""

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

ImportOperation = Literal["new", "update", "delete", "error", "noop"]


class PatientExcelChange(BaseModel):
    """1 フィールドの差分."""

    model_config = ConfigDict(extra="forbid")

    field: str = Field(description="変更フィールド名 (例: address, name)")
    old_value: Any = Field(default=None, description="既存値. 新規行では null.")
    new_value: Any = Field(default=None, description="新しい値. delete 行では null.")


class PatientExcelImportRow(BaseModel):
    """患者マスタシートの 1 行の処理結果."""

    model_config = ConfigDict(extra="forbid")

    row_number: int = Field(
        description="Excel の行番号 (1-indexed, ヘッダーが 1 行目, データは 2 行目から)"
    )
    patient_id: UUID | None = Field(default=None)
    patient_code: str | None = Field(default=None)
    operation: ImportOperation
    changes: list[PatientExcelChange] = Field(default_factory=list)
    error_message: str | None = Field(default=None)


class PfvExcelImportRow(BaseModel):
    """固定訪問スケジュールシートの 1 行の処理結果."""

    model_config = ConfigDict(extra="forbid")

    row_number: int
    patient_id: UUID | None = Field(default=None)
    patient_code: str | None = Field(default=None)
    weekday: int | None = Field(default=None, ge=0, le=6)
    slot_index: int | None = Field(default=None, ge=0, le=1)
    operation: ImportOperation
    changes: list[PatientExcelChange] = Field(default_factory=list)
    error_message: str | None = Field(default=None)


class PatientExcelImportSummary(BaseModel):
    """集計 (UI で件数バッジ表示用)."""

    model_config = ConfigDict(extra="forbid")

    patients_new: int = 0
    patients_update: int = 0
    patients_delete: int = 0
    patients_error: int = 0
    patients_noop: int = 0
    pfv_new: int = 0
    pfv_update: int = 0
    pfv_delete: int = 0
    pfv_error: int = 0
    pfv_noop: int = 0


class PatientExcelImportResponse(BaseModel):
    """POST /api/v1/patients/import-export/import のレスポンス."""

    model_config = ConfigDict(extra="forbid")

    summary: PatientExcelImportSummary
    patient_rows: list[PatientExcelImportRow]
    pfv_rows: list[PfvExcelImportRow]
    transaction_applied: bool = Field(
        description=(
            "True なら DB に反映済 (partial commit). "
            "dry_run=True のときは常に False. "
            "dry_run=False で有効な op が 1 件以上あれば True "
            "(error 行は skip され、それ以外が 1 transaction で commit される). "
            "dry_run=False で有効な op が 0 件 (全 error or 全 noop) なら False."
        )
    )


# ---------------------------------------------------------------------------
# 完全置換 (バックアップ復元) 用 schema
# ---------------------------------------------------------------------------


class PatientExcelReplaceAllSummary(BaseModel):
    """完全置換インポートの集計 (UI で件数バッジ表示用).

    通常 import の summary とは別 schema. PFV は全件再投入する仕様なので
    update 概念は無く、"replace" として件数を返す.
    """

    model_config = ConfigDict(extra="forbid")

    patients_to_create: int = 0
    patients_to_update: int = 0
    patients_to_soft_delete: int = 0  # Excel に無い既存 (alive) 患者
    patients_error: int = 0
    pfv_to_replace: int = 0  # 全件物理削除する既存 PFV の件数
    pfv_to_create: int = 0  # Excel から再投入する PFV の件数
    pfv_error: int = 0


class PatientExcelReplaceAllResponse(BaseModel):
    """POST /api/v1/patients/import-export/replace-all のレスポンス."""

    model_config = ConfigDict(extra="forbid")

    summary: PatientExcelReplaceAllSummary
    patient_rows: list[PatientExcelImportRow]
    pfv_rows: list[PfvExcelImportRow]
    transaction_applied: bool = Field(
        description=(
            "True なら DB に反映済 (atomic). "
            "dry_run=True のときは常に False. "
            "dry_run=False で error が 1 件もなく成功した場合のみ True. "
            "通常 import と違い、error 1 件でも全 rollback されるため True/False "
            "は all-or-nothing."
        )
    )
    patients_to_soft_delete_preview: list[dict] = Field(
        default_factory=list,
        description=(
            "Excel に無い既存 (alive) 患者の一覧 (UI プレビュー表示用). "
            "各要素は {patient_id, patient_code, name} を含む."
        ),
    )
